### ISSUE TITLE
Yolo V5 Inferencing on CPU Machine

### ISSUE THREAD

**Issue by rahul1728jha** [ultralytics/yolov5#10268]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi All,

I am trying to inference using yolo v5. Below is the setup being used:

**System Configuration** : 
Operating system : Windows
Processor	Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz   1.80 GHz
Installed RAM	8.00 GB (7.86 GB usable)
System type	64-bit operating system, x64-based processor

**Setup**:
Input source : Real time RTSP stream from IP camera
Output : Annotated frames

**Code**:
 sample code is like 
```
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True, )
results = detection_model(frame)
labels, confidence, _ = results.xyxyn[0][:, -1], results.xyxyn[0][:, -2], results.xyxyn[0][:, :-1]
detections = results.pred[0]
dets_prod = detections.to("cpu")
```
**Current metrics**
Able to process 30 frames in 6 seconds i.e. getting a output @ **5FPS**

**Question** : 
My Constraint is I have to run it on a CPU.
- My input stream is at 30FPS. With Yolo v5 running on CPU, can i achieve the rate of 30FPS?
- Is it possible to achieve 30FPS on a more powerful machine in terms of RAM and process(No GPU)

**Comment by glenn-jocher** [comment]
@rahul1728jha 👋 Hello! Thanks for asking about **inference speed issues**. I would recommend ONNX or OpenVINO for CPU inference. See https://docs.ultralytics.com/yolov5/tutorials/model_export to ONNX or OpenVINO for up to 3x CPU speedup (https://github.com/ultralytics/yolov5/pull/6613)


PyTorch Hub speeds will vary by hardware, software, model, inference settings, etc. Our default example in Colab with a V100 looks like this:

<img width="1462" alt="Screen Shot 2022-05-03 at 10 20 39 AM" src="https://user-images.githubusercontent.com/26833433/166505744-b7fe9b9a-d2b3-4e95-9441-fa540e0804ef.png">


YOLOv5 🚀 can be run on CPU (i.e. `--device cpu`, slow) or GPU if available (i.e. `--device 0`, faster). You can determine your inference device by viewing the YOLOv5 console output:

### detect.py inference
```bash
python detect.py --weights yolov5s.pt --img 640 --conf 0.25 --source data/images/
```

<img width="1039" alt="Screen Shot 2022-05-03 at 2 48 42 PM" src="https://user-images.githubusercontent.com/26833433/166572240-ae598b34-1084-40f0-a19b-bf00fd3f9050.png">


### YOLOv5 PyTorch [Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) inference
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.print()  # or .show(), .save()
# Speed: 631.5ms pre-process, 19.2ms inference, 1.6ms NMS per image at shape (2, 3, 640, 640)
```

### Increase Speeds

If you would like to increase your inference speed some options are:

- Use batched inference with [YOLOv5 PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
- Reduce `--img-size`, i.e. 1280 -> 640 -> 320
- Reduce model size, i.e. YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s -> YOLOv5n
- Use half precision FP16 inference with `python detect.py --half` and `python val.py --half`
- Use a faster GPUs, i.e.: P100 -> V100 -> A100
- [Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) to ONNX or OpenVINO for up to 3x CPU speedup ([CPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6613))
- [Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) to TensorRT for up to 5x GPU speedup ([GPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6963))
- Use a free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Good luck 🍀 and let us know if you have any other questions!

**Comment by Denizzje** [comment]
I am actually surprised you even got 5fps out of my old laptop CPU lol. I dont think it is possible to run a YoloV5s model with 30fps on your CPU no. 

But you can perhaps squeeze out a few more FPS with the provided tools here: https://docs.ultralytics.com/yolov5/tutorials/model_export

For reference, this is mine on a Ryzen 7 5800X running the benchmark:

```
`benchmarks: weights=yolov5s.pt, imgsz=640, batch_size=1, data=/home/denizz/yolov5/data/coco128.yaml, device=cpu, half=False, test=False, pt_only=False, hard_fail=False
Checking setup...
YOLOv5 🚀 v7.0-4-g7398d2d Python-3.10.4 torch-1.12.0+cu102 CPU
Setup complete ✅ (16 CPUs, 31.3 GB RAM, 77.3/251.0 GB disk)

Benchmarks complete (283.14s)
                   Format  Size (MB)  mAP50-95  Inference time (ms)
0                 PyTorch       14.1    0.4716                59.95
1             TorchScript       28.1    0.4716                58.43
2                    ONNX       28.0    0.4716                43.25
3                OpenVINO        NaN       NaN                  NaN
4                TensorRT        NaN       NaN                  NaN
5                  CoreML        NaN       NaN                  NaN
6   TensorFlow SavedModel       27.8    0.4716                57.90
7     TensorFlow GraphDef       27.8    0.4716                57.96
8         TensorFlow Lite       13.9    0.4714               339.99
9     TensorFlow Edge TPU        NaN       NaN                  NaN
10          TensorFlow.js        NaN       NaN                  NaN
11           PaddlePaddle       56.1    0.4716               526.35`
```

Standard (PyTorch) I get 59.95ms inference time but you can see ONNX is quite a bit faster. I believe it is 17 fps vs 24 fps. OpenVINO fails for me I believe because I have an AMD CPU but it should work on your CPU. Maybe it is faster too. The OPENCV DNN should be the fastest but I couldnt get it to run now because of wrong OpenCV version to compare.

**Comment by lucvt001** [comment]
Hi @glenn-jocher I understand that GPU is blazingly faster than CPU at inferencing on a batch. However, in real applications of inferencing on a live video, does the gpu still have the advantage? Since now we r processing frame by frame? For example, should I get a powerful intel NUC or a Jetson device for my robotic application if I want to do object detection and tracking in real time?

**Comment by glenn-jocher** [comment]
@lucvt001 in real-time applications where inferencing is done on live video frames, GPUs still have a significant advantage over CPUs. While it's true that processing is done frame by frame, the parallel processing capabilities of GPUs allow for faster and more efficient object detection and tracking. 

If you're looking for real-time object detection and tracking in a robotic application, it is recommended to use a powerful GPU such as the one found in an Intel NUC or a Jetson device. These devices are specifically designed for high-performance computing tasks like computer vision and can provide the necessary processing power to achieve real-time performance.

**Comment by lucvt001** [comment]
@glenn-jocher thank you for your advice!

**Comment by glenn-jocher** [comment]
Hi @lucvt001 👋

When it comes to real-time video inferencing for object detection and tracking, GPUs still offer a significant advantage over CPUs. This is due to their parallel processing capabilities, which allow for faster and more efficient computation.

If you're looking to perform real-time object detection and tracking in a robotic application, it is recommended to use a powerful GPU such as those found in Intel NUC or Jetson devices. These devices are specifically designed for high-performance computing tasks like computer vision and can provide the necessary processing power to achieve real-time performance.

I hope this helps! Let me know if you have any more questions.

**Comment by shanmugamani1023** [comment]
thank you

**Comment by glenn-jocher** [comment]
@shanmugamani1023 you're welcome! If you have any more questions in the future, feel free to ask. Good luck with your robotic application! 👍
